### PR TITLE
feat(file source): favor older files and allow configuring greedier reads

### PR DESCRIPTION
**Pull request by lukesteensen**
This is a small experiment for some tweaks that could help the case discussed in #763. It tries to be as noninvasive as possible while still hopefully addressing that use case. 

To recap, the theorized issue is with our "fair" file reading strategy. Right now, we read up to 2kb at a time from each file we're watching, iterating over them in a ~stable but effectively random order (`HashMap::iter`). This prevents one busy file from depriving others of read capacity, but is the opposite of what we want in certain file rotation setups where later files are simply created next to older files. In those cases, we want to prioritize the older file and it's actually better if we ignore the new file until the old one has been processed.

There are three parts to the strategy here, but I've only implemented two so far:

1. Maintain some notion of file age and read from the oldest files first. In this case, we switch from using a `HashMap` to using an `IndexMap` that preserves insertion order. This means that new files that appear while we're running will be read from after files that already existed. It's rough and doesn't account for file age for those that exist at startup, but it's a very minimal change.

2. Expose `max_read_bytes` to the user as an option. This determines how much we read from a single file before moving on to another. It was previously fixed at 2kb, but allowing users to increase it will let us spend more time reading from a single file before continuing to iterate through the rest.

3. TODO: Limit `global_bytes_read` per loop. This would introduce a conditional `break` in the loop over readable files based on the total amount read so far. This would mean that files later in the iteration order could get skipped if earlier files use up the entire read budget. If that limit is equal to `max_read_bytes`, we would exclusively read from the "oldest" file until we catch up (i.e. reach EOF).

The end result is that `max_read_bytes`/`global_bytes_read_limit` essentially gives you a "fairness" ratio and determines the degree to which older files can dominate read capacity over new ones. Using `IndexMap` instead of actually looking at timestamps makes it less precise but it's simpler and still pretty good. We could even check timestamps at startup and do the initial insertion in order if we thought that was worthwhile.

/cc @marcbowes @binarylogic

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice work. Happy to chime in.

>  This prevents one busy file from depriving others of read capacity, but is the opposite of what we want in certain file rotation setups where later files are simply created next to older files. In those cases, we want to prioritize the older file and it's actually better if we ignore the new file until the old one has been processed.

Agree, although, I'd argue this is probably a better default since most logging setups will implement rotation of some sort. Agree?

> We could even check timestamps at startup and do the initial insertion in order if we thought that was worthwhile.

I think this is worthwhile to implement.

> The end result is that max_read_bytes/global_bytes_read_limit essentially gives you a "fairness" ratio and determines the degree to which older files can dominate read capacity over new ones. 

I like this, it is very flexible. Although, I don't know if it's clear to the user how adjusting these settings would actually work without having to first understand the read strategy. We either:

1. Need to document this well
2. Or, consider a generic `greedy_read` boolean, or something more generic.

Does that make sense?

---

To add to this, I could see scenarios where a user might prefer new data first, instead of old data, and I assume it would be trivial to add a setting for that?

**Comment by lukesteensen** [comment]
> I'd argue this is probably a better default since most logging setups will implement rotation of some sort. Agree?

I'm not sure it really matters in most cases. Normally, we should be keeping up with the end of the current log file as it's being written to. So when the file gets rotated, our reads can switch pretty much at the same time as the write stream. This really only comes into play when there are multiple "active" files, where it might be desirable to read from them fairly (e.g. nginx `access.log` and `error.log`).

I did just notice our docs say "Ensure the `includes` paths include rotated files", so we may be on different pages here. I would not expect rotated files to be included because they're not "active" tailing targets. The only reason I can think to include them would be to backfill historical data.

Ultimate, it seems like this boils down to which situation we would rather handle gracefully by default:
1. Falling behind on a rotated file
2. One busy file starving separate files of read capacity

Handling (2) seems good, but there is a good argument I think for (1) being far more common. What do you think?

---

>     2. consider a generic `greedy_read` boolean, or something more generic.

Agreed, my current iteration simplifies this down to an `oldest_first` flag that causes the file server not to iterate past the oldest file that we get a "full" read out of. We can probably come up with a better name.

**Comment by lukesteensen** [comment]
Alright, I think this is ready for review. Quick summary of the changes:

1. Use insertion-order-preserving `IndexMap` to store file entries
2. At startup, do an initial glob lookup of existing files and add them to the map in order of creation time
3. While looping over the map of files and reading lines, skip the remaining (i.e. younger) files if we ever hit a "max read" (indicating that we are not yet up to date with that file)

I don't love the name `oldest_first`, so am curious if others have good ideas for this.

I'm also on the fence about exposing `max_read_bytes`. It's not something I'd recommend people mess with, but it is useful for testing and pretty harmless? I can find a way to hide it if people would prefer.

**Comment by binarylogic** [comment]
@lukesteensen can you add a section on why someone would use this? Ex: "Greedy vs Non-greedy Reads"

File: .metadata.toml [hunk 1]
```diff
@@ -189,6 +189,28 @@ will buffer lines into a single event before flushing, regardless of whether \
...
+[sources.file.options.oldest_first]
+type = "bool"
+default = false
+null = true
+description = """\
+By default, the file source will try to share read capacity fairly between all \
+the files that it is currently tailing. Depending on your file rotation \
+strategy, log file organization, and throughput patterns, it could be better \
+for the file source to focus on draining the oldest files it's aware of before \
+moving on to read data from younger files. This flag enables that behavior.\
```
Comment (LucioFranco) [inline .metadata.toml]:
```suggestion
moving on to read data from younger files. This flag enables prioritizing older files.\
```

File: lib/file-source/src/file_server.rs [hunk 2]
```diff
@@ -202,13 +232,18 @@ impl FileServer {
...
                         break;
                     }
                 }
                 if bytes_read > 0 {
                     global_bytes_read = global_bytes_read.saturating_add(bytes_read);
                     checkpointer.set_checkpoint(file_id, watcher.get_file_position());
                 }
+                // Do not move on to newer files if we are behind on an older file
+                if self.oldest_first && maxed_out_reading_single_file {
+                    break;
```
Comment (LucioFranco) [inline lib/file-source/src/file_server.rs]: I'm curious if there is a cleaner way to express this double loop its starting to get a bit hard to follow :(

File: lib/file-source/src/file_server.rs [hunk 3]
```diff
@@ -248,6 +283,30 @@ impl FileServer {
...
+        fp_map: &mut IndexMap<FileFingerprint, FileWatcher>,
+        checkpointer: &Checkpointer,
+        read_from_beginning: bool,
+    ) {
+        let file_position = if read_from_beginning {
+            0
+        } else {
+            checkpointer.get_checkpoint(file_id).unwrap_or(0)
+        };
+        if let Ok(mut watcher) = FileWatcher::new(path, file_position, self.ignore_before) {
```
Comment (LucioFranco) [inline lib/file-source/src/file_server.rs]: should probably print an error here?

**Review by LucioFranco — APPROVED** [review]
LGTM just left some questions inline

File: lib/file-source/src/file_server.rs (hunk 2 above)
Comment (lukesteensen) [reply lib/file-source/src/file_server.rs]: There are some general refactoring I've been thinking about and I might take a quick stab at those after getting this merged.

Comment (LucioFranco) [reply lib/file-source/src/file_server.rs]: 👍 sounds good to me

File: lib/file-source/src/file_server.rs (hunk 3 above)
Comment (lukesteensen) [reply lib/file-source/src/file_server.rs]: This was just moved, but yeah good idea. I'll add it in a followup.

**Comment by binarylogic** [comment]
Thanks @lukesteensen, looks good! I think this is good for merging.
